// md-editor-v3/lib/es/MdEditor.mjs
import { a as te, D as at } from "./chunks/index.mjs";
import { s as lt, a as nt, M as We } from "./chunks/index4.mjs";
import { C as rt, c as st, a as it, e as ct, b as ut, u as dt, d as mt, f as bt, g as vt, h as gt, i as ht, j as ft, k as pt, M as Tt } from "./chunks/index3.mjs";
import wt from "./ModalToolbar.mjs";
import yt from "./NormalToolbar.mjs";
import { inject as r, watch as N, nextTick as ie, onMounted as K, ref as I, defineComponent as C, createVNode as a, reactive as Y, toRef as ke, computed as P, Fragment as _e, cloneVNode as ne, createApp as Ct, shallowRef as kt, onBeforeUnmount as he } from "vue";
import { g as Z, p as l, a as $t, M as ae, d as xt, b as St } from "./chunks/config.mjs";
import { linkTo as It, throttle as Et, createSmoothScroll as ze, draggingScroll as Lt } from "@vavt/util";
import { deleteLine as Nt, history as $e, defaultKeymap as At, historyKeymap as Dt, indentWithTab as Rt, undo as Mt, redo as Ft } from "@codemirror/commands";
import { markdown as Bt } from "@codemirror/lang-markdown";
import { languages as Pt } from "@codemirror/language-data";
import { EditorState as ce, EditorSelection as U, Compartment as j, StateEffect as re, StateField as qe, RangeSetBuilder as Ht } from "@codemirror/state";
import { placeholder as Ot, EditorView as B, showTooltip as Vt, Decoration as Wt, WidgetType as _t, keymap as zt, drawSelection as qt } from "@codemirror/view";
import { b as h, E as X, R as y, C as Ut, U as fe, a as Ue, c as Ke, O as Ge, d as Kt, e as Gt, T as jt, S as Yt, G as xe } from "./chunks/event-bus.mjs";
import { indentUnit as Zt, syntaxHighlighting as je, HighlightStyle as Ye } from "@codemirror/language";
import { autocompletion as Xt } from "@codemirror/autocomplete";
import { a as Jt } from "./chunks/dom.mjs";
import { I as x, M as Qt } from "./chunks/index2.mjs";
import { b as eo } from "./chunks/index5.mjs";
import { tags as b } from "@lezer/highlight";
import { searchKeymap as to } from "@codemirror/search";
import { g as Se } from "./chunks/vue-tsx.mjs";
const oo = (e, t, o) => {
  const n = r("editorId"), i = r("setting");
  let s = () => {
  }, c = () => {
  };
  const d = () => {
    s();
    const u = o.value?.view.contentDOM.getRootNode(), v = u?.querySelector(
      `#${n} .cm-scroller`
    ), m = u?.querySelector(
      `[id="${n}-preview-wrapper"]`
    ), f = u?.querySelector(
      `[id="${n}-html-wrapper"]`
    );
    (m || f) && ([c, s] = (m ? lt : nt)(
      v,
      m || f,
      o.value
    ), e.scrollAuto && c());
  };
  N([t, i], () => {
    ie(d);
  }), N(
    () => e.scrollAuto,
    (u) => {
      u ? c() : s();
    }
  ), N(
    () => i.value.previewOnly,
    (u) => {
      u ? s() : c();
    }
  ), K(d);
}, ue = async (e, t, o) => {
  if (/^h[1-6]$/.test(e))
    return ao(e, t);
  if (e === "prettier")
    return await lo(t, o);
  switch (e) {
    case "bold":
    case "underline":
    case "italic":
    case "strikeThrough":
    case "sub":
    case "sup":
    case "codeRow":
    case "katexInline":
    case "katexBlock":
      return ro(e, t);
    case "quote":
    case "orderedList":
    case "unorderedList":
    case "task":
      return io(e, t);
    case "code":
      return co(o, t);
    case "table":
      return bo(o);
    case "link": {
      const n = t.getSelectedText(), { desc: i = n, url: s = "" } = o, c = `[${i}](${s})`;
      return {
        text: c,
        options: {
          select: s === "",
          deviationStart: c.length - s.length - 1,
          deviationEnd: -1
        }
      };
    }
    case "image":
      return mo(o, t);
    case "flow":
    case "sequence":
    case "gantt":
    case "class":
    case "state":
    case "pie":
    case "relationship":
    case "journey":
      return uo(e);
    case "universal":
      return vo(t.getSelectedText(), o);
    default:
      return { text: "", options: {} };
  }
}, ao = (e, t) => {
  const o = e.slice(1), n = "#".repeat(Number(o)), [i, s, c] = pe(t, {
    wholeLine: !0
  });
  return {
    text: `${n} ${i}`,
    options: { deviationStart: n.length + 1, replaceStart: s, replaceEnd: c }
  };
}, lo = async (e, t) => {
  const o = window.prettier || Z.editorExtensions.prettier?.prettierInstance, n = [
    window.prettierPlugins?.markdown || Z.editorExtensions.prettier?.parserMarkdownInstance
  ];
  return !o || !n[0] ? (h.emit(t.editorId, X, {
    name: "prettier",
    message: "prettier is undefined"
  }), {
    text: e.getValue(),
    options: { select: !1, replaceAll: !0 }
  }) : {
    text: await o.format(e.getValue(), {
      parser: "markdown",
      plugins: n
    }),
    options: { select: !1, replaceAll: !0 }
  };
}, no = {
  bold: ["**", "**", 2, -2],
  underline: ["<u>", "</u>", 3, -4],
  italic: ["*", "*", 1, -1],
  strikeThrough: ["~~", "~~", 2, -2],
  sub: ["~", "~", 1, -1],
  sup: ["^", "^", 1, -1],
  codeRow: ["`", "`", 1, -1],
  katexInline: ["$", "$", 1, -1],
  katexBlock: [`
$$
`, `
$$
`, 4, -4]
}, ro = (e, t) => {
  const o = t.getSelectedText(), [n, i, s, c] = no[e];
  return {
    text: `${n}${o}${i}`,
    options: {
      deviationStart: s,
      deviationEnd: c
      // replaceStart, replaceEnd
    }
  };
}, so = {
  quote: "> ",
  unorderedList: "- ",
  orderedList: 1,
  task: "- [ ] "
}, io = (e, t) => {
  const [o, n, i] = pe(t, {
    wholeLine: !0
  }), s = o.split(`
`), c = so[e], d = e === "orderedList" ? s.map((m, f) => `${c + f}. ${m}`) : s.map((m) => `${c}${m}`), u = e === "orderedList" ? "1. " : c.toString(), v = s.length === 1 ? u.length : 0;
  return {
    text: d.join(`
`),
    options: { deviationStart: v, replaceStart: n, replaceEnd: i }
  };
}, co = (e, t) => {
  const [o, n, i] = pe(t), s = e.mode || "language", c = `
\`\`\`${s}
${e.text || o || ""}
\`\`\`
`;
  return {
    text: c,
    options: {
      deviationStart: 4,
      deviationEnd: 4 + s.length - c.length,
      replaceStart: n,
      replaceEnd: i
    }
  };
}, uo = (e) => ({
  text: `
\`\`\`mermaid
${{
    flow: `flowchart TD 
  Start --> Stop`,
    sequence: `sequenceDiagram
  A->>B: hello!
  B-->>A: hi!`,
    gantt: `gantt
title Gantt Chart
dateFormat  YYYY-MM-DD`,
    class: `classDiagram
  class Animal`,
    state: `stateDiagram-v2
  s1 --> s2`,
    pie: `pie
  "Dogs" : 386
  "Cats" : 85
  "Rats" : 15`,
    relationship: `erDiagram
  CAR ||--o{ NAMED-DRIVER : allows`,
    journey: `journey
  title My Journey`,
    ...Z.editorConfig.mermaidTemplate
  }[e]}
\`\`\`
`,
  options: { deviationStart: 12, deviationEnd: -5 }
}), mo = (e, t) => {
  const o = t.getSelectedText(), { desc: n = o, url: i = "", urls: s } = e;
  let c = "";
  const d = i === "" && (!s || s instanceof Array && s.length === 0);
  return s instanceof Array ? c = s.reduce((u, v) => {
    const {
      url: m = "",
      alt: f = "",
      title: g = ""
    } = typeof v == "object" ? v : { url: v };
    return u + `![${f}](${m}${g ? " '" + g + "'" : ""})
`;
  }, "") : c = `![${n}](${i})
`, {
    text: c,
    options: {
      select: i === "",
      deviationStart: d ? c.length - i.length - 2 : c.length,
      deviationEnd: d ? -2 : 0
    }
  };
}, bo = (e) => {
  const { selectedShape: t = { x: 1, y: 1 } } = e, { x: o, y: n } = t;
  let i = `
| Column`;
  for (let s = 0; s <= n; s++)
    i += " |";
  i += `
|`;
  for (let s = 0; s <= n; s++)
    i += " - |";
  for (let s = 0; s <= o; s++) {
    i += `
|`;
    for (let c = 0; c <= n; c++)
      i += " |";
  }
  return i += `
`, { text: i, options: { deviationStart: 3, deviationEnd: 10 - i.length } };
}, vo = (e, t) => {
  const { generate: o } = t, n = o(e);
  return {
    text: n.targetValue,
    options: {
      select: n.select ?? !0,
      deviationStart: n.deviationStart || 0,
      deviationEnd: n.deviationEnd || 0
    }
  };
}, pe = (e, t = { wholeLine: !1 }) => {
  const o = e.view.state, n = o.selection.main;
  if (n.empty) {
    const i = o.doc.lineAt(n.from);
    return [o.doc.lineAt(n.from).text, i.from, i.to];
  } else if (t.wholeLine) {
    const i = o.doc.lineAt(n.from), s = o.doc.lineAt(n.to);
    return [
      o.doc.sliceString(i.from, s.to),
      i.from,
      s.to
    ];
  }
  return [
    o.doc.sliceString(n.from, n.to),
    n.from,
    n.to
  ];
}, G = (e) => {
  const t = new j();
  return (n) => (t.get(e.state) ? e.dispatch({ effects: t.reconfigure(n) }) : e.dispatch({
    effects: re.appendConfig.of(t.of(n))
  }), !0);
};
class go {
  view;
  maxLength = Number.MAX_SAFE_INTEGER;
  // 切换tabSize的执行方法。切换时，Compartment实例需要相同
  toggleTabSize;
  togglePlaceholder;
  /**
   * 设置全部的扩展
   */
  setExtensions;
  toggleDisabled;
  toggleReadOnly;
  toggleMaxlength;
  getValue() {
    return this.view.state.doc.toString();
  }
  /**
   * 设置内容
   *
   * @param insert 待插入内容
   * @param from 插入开始位置
   * @param to 插入结束位置
   */
  setValue(t, o = 0, n = this.view.state.doc.length) {
    this.view.dispatch({
      changes: {
        from: o,
        to: n,
        insert: t
      }
    });
  }
  /**
   * 获取选中的文本
   */
  getSelectedText() {
    const { from: t, to: o } = this.view.state.selection.main;
    return this.view.state.sliceDoc(t, o);
  }
  /**
   * 使用新的内容替换选中的内容
   *
   * @param text 待替换内容
   * @param options 替换后是否选中
   */
  replaceSelectedText(t, o, n) {
    const i = {
      // 是否选中
      select: !0,
      // 选中时，开始位置的偏移量
      deviationStart: 0,
      // 结束的偏移量
      deviationEnd: 0,
      // 直接替换所有文本
      replaceAll: !1,
      // 替换旧文本的开始位置
      replaceStart: -1,
      // 替换旧文本的结束位置
      replaceEnd: -1,
      ...o
    };
    try {
      if (i.replaceAll) {
        if (this.setValue(t), t.length > this.maxLength)
          throw new Error("The input text is too long");
        return;
      }
      if (this.view.state.doc.length - this.getSelectedText().length + t.length > this.maxLength)
        throw new Error("The input text is too long");
      const { from: s } = this.view.state.selection.main;
      i.replaceStart !== -1 ? this.view.dispatch({
        changes: {
          from: i.replaceStart,
          to: i.replaceEnd,
          insert: t
        }
      }) : this.view.dispatch(this.view.state.replaceSelection(t)), i.select && this.view.dispatch({
        selection: {
          anchor: i.replaceStart === -1 ? s + i.deviationStart : i.replaceStart + i.deviationStart,
          head: i.replaceStart === -1 ? s + t.length + i.deviationEnd : i.replaceStart + t.length + i.deviationEnd
        }
      }), this.view.focus();
    } catch (s) {
      if (s.message === "The input text is too long")
        h.emit(n, X, {
          name: "overlength",
          message: s.message,
          data: t
        });
      else
        throw s;
    }
  }
  constructor(t) {
    this.view = t, this.toggleTabSize = G(this.view), this.togglePlaceholder = G(this.view), this.setExtensions = G(this.view), this.toggleDisabled = G(this.view), this.toggleReadOnly = G(this.view), this.toggleMaxlength = G(this.view);
  }
  /**
   * 设置tabSize
   *
   * @param tabSize 需要切换的大小
   */
  setTabSize(t) {
    this.toggleTabSize([
      ce.tabSize.of(t),
      Zt.of(" ".repeat(t))
    ]);
  }
  /**
   * 设置placeholder
   *
   * @param t 目标内容
   */
  setPlaceholder(t) {
    this.togglePlaceholder(Ot(t));
  }
  focus(t) {
    if (this.view.focus(), !t)
      return;
    let o = 0, n = 0, i = 0;
    switch (t) {
      case "start":
        break;
      case "end": {
        o = n = i = this.getValue().length;
        break;
      }
      default:
        o = t.rangeAnchor || t.cursorPos, n = t.rangeHead || t.cursorPos, i = t.cursorPos;
    }
    this.view.dispatch({
      scrollIntoView: !0,
      selection: U.create(
        [U.range(o, n), U.cursor(i)],
        1
      )
    });
  }
  setDisabled(t) {
    this.toggleDisabled([B.editable.of(!t)]);
  }
  setReadOnly(t) {
    this.toggleReadOnly([ce.readOnly.of(t)]);
  }
  setMaxLength(t) {
    this.maxLength = t, this.toggleMaxlength([
      ce.changeFilter.of((o) => o.newDoc.length <= t)
    ]);
  }
}
const ho = (e, t) => {
  if (e === t) return !0;
  if (e.length !== t.length) return !1;
  for (let o = 0; o < e.length; o++)
    if (e[o] !== t[o]) return !1;
  return !0;
}, fo = (e, t) => {
  const o = I(t.value);
  N([t], () => {
    (!o.value || !ho(o.value, t.value)) && (o.value = t.value, e());
  });
}, le = (e, t, o, n, i) => (s, c, d, u) => {
  const v = `${e}${t}${o}${n}`, m = d + c.label.length + (i === "title" ? o.length : 0);
  s.dispatch({
    changes: { from: d, to: u, insert: v },
    selection: U.create(
      [
        U.range(
          d + c.label.length + (i === "title" ? 1 : -t.length),
          m
        ),
        U.cursor(m)
      ],
      1
    )
  }), s.focus();
}, Ie = (e) => (t, o, n, i) => {
  const s = e.slice(i - n);
  t.dispatch(t.state.replaceSelection(`${s} `));
}, Ee = (e) => {
  const t = (o) => {
    const n = o.matchBefore(
      /^#+|^-\s*\[*\s*\]*|`+|\[|!\[*|^\|\s?\|?|\$\$?|!+\s*\w*/
    );
    return n === null || n.from == n.to && o.explicit ? null : {
      from: n.from,
      options: [
        // 标题
        ...["h2", "h3", "h4", "h5", "h6"].map((i, s) => {
          const c = new Array(s + 2).fill("#").join("");
          return {
            label: c,
            type: "text",
            apply: Ie(c)
          };
        }),
        // 任务列表
        ...["unchecked", "checked"].map((i) => {
          const s = i === "checked" ? "- [x]" : "- [ ]";
          return {
            label: s,
            type: "text",
            apply: Ie(s)
          };
        }),
        // 代码
        ...[
          ["`", ""],
          ["```", "language"],
          ["```mermaid\n", ""],
          ["```echarts\n", ""]
        ].map((i) => ({
          label: `${i[0]}${i[1]}`,
          type: "text",
          apply: le(i[0], i[1], "", i[0] === "`" ? "`" : "\n```", "type")
        })),
        // 链接
        {
          label: "[]()",
          type: "text"
        },
        {
          label: "![]()",
          type: "text"
        },
        // 表格
        {
          label: "| |",
          type: "text",
          detail: "table",
          apply: `| col | col | col |
| - | - | - |
| content | content | content |
| content | content | content |`
        },
        // 公式
        {
          label: "$",
          type: "text",
          apply: le("$", "", "", "$", "type")
        },
        {
          label: "$$",
          type: "text",
          apply: le("$$", "", `
`, `
$$`, "title")
        },
        // 那啥？
        ...[
          "note",
          "abstract",
          "info",
          "tip",
          "success",
          "question",
          "warning",
          "failure",
          "danger",
          "bug",
          "example",
          "quote",
          "hint",
          "caution",
          "error",
          "attention"
        ].map((i) => ({
          label: `!!! ${i}`,
          type: "text",
          apply: le("!!!", ` ${i}`, " Title", `

!!!`, "title")
        }))
      ]
    };
  };
  return Xt({
    override: e ? [t, ...e] : [t]
  });
}, po = /* @__PURE__ */ C({
  name: `${l}-divider`,
  setup() {
    return () => a("div", {
      class: `${l}-divider`
    }, null);
  }
}), To = /* @__PURE__ */ C({
  name: "ToolbarBold",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.bold,
      "aria-label": t.value.toolbarTips?.bold,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "bold");
      },
      type: "button"
    }, [a(x, {
      name: "bold"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.bold])]);
  }
}), wo = /* @__PURE__ */ C({
  name: "ToolbarCatalog",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = r("catalogVisible");
    return () => a("button", {
      class: [`${l}-toolbar-item`, i.value && `${l}-toolbar-active`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.catalog,
      "aria-label": t.value.toolbarTips?.catalog,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, Ut);
      },
      key: "bar-catalog",
      type: "button"
    }, [a(x, {
      name: "catalog"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.catalog])]);
  }
}), yo = /* @__PURE__ */ C({
  name: "ToolbarCode",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.code,
      "aria-label": t.value.toolbarTips?.code,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "code");
      },
      type: "button"
    }, [a(x, {
      name: "code"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.code])]);
  }
}), Co = /* @__PURE__ */ C({
  name: "ToolbarCodeRow",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.codeRow,
      "aria-label": t.value.toolbarTips?.codeRow,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "codeRow");
      },
      type: "button"
    }, [a(x, {
      name: "code-row"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.codeRow])]);
  }
}), ko = /* @__PURE__ */ C({
  name: "ToolbarFullscreen",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), {
      fullscreenHandler: i
    } = ta();
    return () => a("button", {
      class: [`${l}-toolbar-item`, n.value.fullscreen && `${l}-toolbar-active`, t?.value && `${l}-disabled`],
      title: e.value.toolbarTips?.fullscreen,
      "aria-label": e.value.toolbarTips?.fullscreen,
      disabled: t?.value,
      onClick: () => {
        i();
      },
      type: "button"
    }, [a(x, {
      name: n.value.fullscreen ? "fullscreen-exit" : "fullscreen"
    }, null), o?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [e.value.toolbarTips?.fullscreen])]);
  }
}), $o = /* @__PURE__ */ C({
  name: "ToolbarGithub",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, t?.value && `${l}-disabled`],
      title: e.value.toolbarTips?.github,
      "aria-label": e.value.toolbarTips?.github,
      disabled: t?.value,
      onClick: () => {
        It("https://github.com/imzbf/md-editor-v3");
      },
      type: "button"
    }, [a(x, {
      name: "github"
    }, null), o?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [e.value.toolbarTips?.github])]);
  }
}), xo = /* @__PURE__ */ C({
  name: "ToolbarHtmlPreview",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => a("button", {
      class: [`${l}-toolbar-item`, n.value.htmlPreview && `${l}-toolbar-active`, t?.value && `${l}-disabled`],
      title: e.value.toolbarTips?.htmlPreview,
      "aria-label": e.value.toolbarTips?.htmlPreview,
      disabled: t?.value,
      onClick: () => {
        i("htmlPreview");
      },
      type: "button"
    }, [a(x, {
      name: "preview-html"
    }, null), o?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [e.value.toolbarTips?.htmlPreview])]);
  }
}), So = /* @__PURE__ */ C({
  name: "ToolbarImage",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.image,
      "aria-label": t.value.toolbarTips?.image,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "image");
      },
      type: "button"
    }, [a(x, {
      name: "image"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.image])]);
  }
}), Io = {
  visible: {
    type: Boolean,
    default: !1
  },
  onCancel: {
    type: Function,
    default: () => {
    }
  },
  onOk: {
    type: Function,
    default: () => {
    }
  }
}, Eo = /* @__PURE__ */ C({
  name: `${l}-modal-clip`,
  props: Io,
  setup(e) {
    const t = r("usedLanguageText"), o = r("editorId"), n = r("rootRef");
    let i = Z.editorExtensions.cropper.instance;
    const s = I(), c = I(), d = I(), u = Y({
      cropperInited: !1,
      imgSelected: !1,
      imgSrc: "",
      // 是否全屏
      isFullscreen: !1
    });
    let v = null;
    N(() => e.visible, () => {
      e.visible && !u.cropperInited && (i = i || window.Cropper, s.value.onchange = () => {
        if (!i) {
          h.emit(o, X, {
            name: "Cropper",
            message: "Cropper is undefined"
          });
          return;
        }
        const g = s.value.files || [];
        if (u.imgSelected = !0, g?.length > 0) {
          const p = new FileReader();
          p.onload = (k) => {
            u.imgSrc = k.target.result;
          }, p.readAsDataURL(g[0]);
        }
      });
    }), N(() => [u.imgSelected], () => {
      d.value.style = "";
    }), N([ke(() => u.isFullscreen), ke(() => u.imgSrc)], () => {
      u.imgSrc && ie(() => {
        v?.destroy(), d.value.style = "", c.value && (v = new i(c.value, {
          viewMode: 2,
          preview: n.value.getRootNode().querySelector(`.${l}-clip-preview-target`)
          // aspectRatio: 16 / 9,
        }));
      });
    });
    const m = () => {
      v.clear(), v.destroy(), v = null, s.value.value = "", u.imgSelected = !1, u.imgSrc = "";
    }, f = (g) => {
      u.isFullscreen = g;
    };
    return () => a(Qt, {
      class: `${l}-modal-clip`,
      title: t.value.clipModalTips?.title,
      visible: e.visible,
      onClose: e.onCancel,
      showAdjust: !0,
      isFullscreen: u.isFullscreen,
      onAdjust: f,
      width: "668px",
      height: "421px"
    }, {
      default: () => [a("div", {
        class: `${l}-form-item ${l}-clip`
      }, [a("div", {
        class: `${l}-clip-main`
      }, [u.imgSelected ? a("div", {
        class: `${l}-clip-cropper`
      }, [a("img", {
        src: u.imgSrc,
        ref: c,
        style: {
          display: "none"
        },
        alt: ""
      }, null), a("div", {
        class: `${l}-clip-delete`,
        onClick: m
      }, [a(x, {
        name: "delete"
      }, null)])]) : a("div", {
        class: `${l}-clip-upload`,
        onClick: () => {
          s.value.click();
        },
        role: "button",
        tabindex: "0",
        "aria-label": t.value.imgTitleItem?.upload
      }, [a(x, {
        name: "upload"
      }, null)])]), a("div", {
        class: `${l}-clip-preview`
      }, [a("div", {
        class: `${l}-clip-preview-target`,
        ref: d
      }, null)])]), a("div", {
        class: `${l}-form-item`
      }, [a("button", {
        class: `${l}-btn`,
        type: "button",
        onClick: () => {
          if (v) {
            const g = v.getCroppedCanvas();
            h.emit(o, fe, [eo(g.toDataURL("image/png"))], e.onOk), m();
          }
        }
      }, [t.value.clipModalTips?.buttonUpload || t.value.linkModalTips?.buttonOK])]), a("input", {
        ref: s,
        accept: "image/*",
        type: "file",
        multiple: !1,
        style: {
          display: "none"
        },
        "aria-hidden": "true"
      }, null)]
    });
  }
}), Lo = {
  clipVisible: {
    type: Boolean,
    default: !1
  },
  onCancel: {
    type: Function,
    default: () => {
    }
  },
  onOk: {
    type: Function,
    default: () => {
    }
  }
}, No = /* @__PURE__ */ C({
  name: `${l}-modals`,
  props: Lo,
  setup(e) {
    return () => a(Eo, {
      visible: e.clipVisible,
      onOk: e.onOk,
      onCancel: e.onCancel
    }, null);
  }
}), Ao = /* @__PURE__ */ C({
  name: "ToolbarImageDropdown",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = I(!1), c = I(!1), d = I(), u = () => {
      h.emit(e, fe, Array.from(d.value.files || [])), d.value.value = "";
    }, v = (k, E) => {
      o?.value || h.emit(e, y, k, E);
    };
    K(() => {
      d.value.addEventListener("change", u);
    });
    const m = (k) => {
      s.value = k;
    }, f = () => {
      c.value = !1;
    }, g = (k) => {
      k && v("image", {
        desc: k.desc,
        url: k.url,
        transform: !0
      }), c.value = !1;
    }, p = P(() => a("ul", {
      class: `${l}-menu`,
      onClick: () => {
        s.value = !1;
      },
      role: "menu"
    }, [a("li", {
      class: `${l}-menu-item ${l}-menu-item-image`,
      onClick: () => {
        v("image");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.imgTitleItem?.link]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-image`,
      onClick: () => {
        d.value.click();
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.imgTitleItem?.upload]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-image`,
      onClick: () => {
        c.value = !0;
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.imgTitleItem?.clip2upload])]));
    return () => a(_e, null, [a("label", {
      for: `${i}_label`,
      style: {
        display: "none"
      },
      "aria-label": t.value.imgTitleItem?.upload
    }, null), a("input", {
      id: `${i}_label`,
      ref: d,
      accept: "image/*",
      type: "file",
      multiple: !0,
      style: {
        display: "none"
      }
    }, null), a(te, {
      relative: `#${i}`,
      visible: s.value,
      onChange: m,
      disabled: o?.value,
      overlay: p.value
    }, {
      default: () => [a("button", {
        class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
        title: t.value.toolbarTips?.image,
        "aria-label": t.value.toolbarTips?.image,
        disabled: o?.value,
        type: "button"
      }, [a(x, {
        name: "image"
      }, null), n?.value && a("div", {
        class: `${l}-toolbar-item-name`
      }, [t.value.toolbarTips?.image])])]
    }), a(No, {
      clipVisible: c.value,
      onCancel: f,
      onOk: g
    }, null)]);
  }
}), Do = /* @__PURE__ */ C({
  name: "ToolbarItalic",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.italic,
      "aria-label": t.value.toolbarTips?.italic,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "italic");
      },
      type: "button"
    }, [a(x, {
      name: "italic"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.italic])]);
  }
}), Ro = /* @__PURE__ */ C({
  name: "ToolbarKatex",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = I(!1), c = (v) => {
      o?.value || h.emit(e, y, v);
    }, d = (v) => {
      s.value = v;
    }, u = P(() => a("ul", {
      class: `${l}-menu`,
      onClick: () => {
        s.value = !1;
      },
      role: "menu"
    }, [a("li", {
      class: `${l}-menu-item ${l}-menu-item-katex`,
      onClick: () => {
        c("katexInline");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.katex?.inline]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-katex`,
      onClick: () => {
        c("katexBlock");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.katex?.block])]));
    return () => a(te, {
      relative: `#${i}`,
      visible: s.value,
      onChange: d,
      disabled: o?.value,
      overlay: u.value,
      key: "bar-katex"
    }, {
      default: () => [a("button", {
        class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
        title: t.value.toolbarTips?.katex,
        "aria-label": t.value.toolbarTips?.katex,
        disabled: o?.value,
        type: "button"
      }, [a(x, {
        name: "formula"
      }, null), n?.value && a("div", {
        class: `${l}-toolbar-item-name`
      }, [t.value.toolbarTips?.katex])])]
    });
  }
}), Mo = /* @__PURE__ */ C({
  name: "ToolbarLink",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.link,
      "aria-label": t.value.toolbarTips?.link,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "link");
      },
      type: "button"
    }, [a(x, {
      name: "link"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.link])]);
  }
}), Fo = /* @__PURE__ */ C({
  name: "ToolbarMermaid",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = I(!1), c = (v) => {
      o?.value || h.emit(e, y, v);
    }, d = (v) => {
      s.value = v;
    }, u = P(() => a("ul", {
      class: `${l}-menu`,
      onClick: () => {
        s.value = !1;
      },
      role: "menu"
    }, [a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("flow");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.flow]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("sequence");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.sequence]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("gantt");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.gantt]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("class");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.class]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("state");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.state]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("pie");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.pie]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("relationship");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.relationship]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-mermaid`,
      onClick: () => {
        c("journey");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.mermaid?.journey])]));
    return () => a(te, {
      relative: `#${i}`,
      visible: s.value,
      onChange: d,
      disabled: o?.value,
      overlay: u.value,
      key: "bar-mermaid"
    }, {
      default: () => [a("button", {
        class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
        title: t.value.toolbarTips?.mermaid,
        "aria-label": t.value.toolbarTips?.mermaid,
        disabled: o?.value,
        type: "button"
      }, [a(x, {
        name: "mermaid"
      }, null), n?.value && a("div", {
        class: `${l}-toolbar-item-name`
      }, [t.value.toolbarTips?.mermaid])])]
    });
  }
}), Bo = /* @__PURE__ */ C({
  name: "ToolbarNext",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.next,
      "aria-label": t.value.toolbarTips?.next,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, Ue);
      },
      type: "button"
    }, [a(x, {
      name: "next"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.next])]);
  }
}), Po = /* @__PURE__ */ C({
  name: "ToolbarOrderedList",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.orderedList,
      "aria-label": t.value.toolbarTips?.orderedList,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "orderedList");
      },
      type: "button"
    }, [a(x, {
      name: "ordered-list"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.orderedList])]);
  }
}), Ho = /* @__PURE__ */ C({
  name: "ToolbarPageFullscreen",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => a("button", {
      class: [`${l}-toolbar-item`, n.value.pageFullscreen && `${l}-toolbar-active`, t?.value && `${l}-disabled`],
      title: e.value.toolbarTips?.pageFullscreen,
      "aria-label": e.value.toolbarTips?.pageFullscreen,
      disabled: t?.value,
      onClick: () => {
        i("pageFullscreen");
      },
      type: "button"
    }, [a(x, {
      name: n.value.pageFullscreen ? "minimize" : "maximize"
    }, null), o?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [e.value.toolbarTips?.pageFullscreen])]);
  }
}), Oo = /* @__PURE__ */ C({
  name: "ToolbarPrettier",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.prettier,
      "aria-label": t.value.toolbarTips?.prettier,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "prettier");
      },
      type: "button"
    }, [a(x, {
      name: "prettier"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.prettier])]);
  }
}), Vo = /* @__PURE__ */ C({
  name: "ToolbarPreview",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => a("button", {
      class: [`${l}-toolbar-item`, n.value.preview && `${l}-toolbar-active`, t?.value && `${l}-disabled`],
      title: e.value.toolbarTips?.preview,
      "aria-label": e.value.toolbarTips?.preview,
      disabled: t?.value,
      onClick: () => {
        i("preview");
      },
      type: "button"
    }, [a(x, {
      name: "preview"
    }, null), o?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [e.value.toolbarTips?.preview])]);
  }
}), Wo = /* @__PURE__ */ C({
  name: "ToolbarPreviewOnly",
  setup() {
    const e = r("usedLanguageText"), t = r("disabled"), o = r("showToolbarName"), n = r("setting"), i = r("updateSetting");
    return () => a("button", {
      class: [`${l}-toolbar-item`, n.value.previewOnly && `${l}-toolbar-active`, t?.value && `${l}-disabled`],
      title: e.value.toolbarTips?.previewOnly,
      "aria-label": e.value.toolbarTips?.previewOnly,
      disabled: t?.value,
      onClick: () => {
        i("previewOnly");
      },
      type: "button"
    }, [a(x, {
      name: "preview-only"
    }, null), o?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [e.value.toolbarTips?.previewOnly])]);
  }
}), _o = /* @__PURE__ */ C({
  name: "ToolbarQuote",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.quote,
      "aria-label": t.value.toolbarTips?.quote,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "quote");
      },
      type: "button"
    }, [a(x, {
      name: "quote"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.quote])]);
  }
}), zo = /* @__PURE__ */ C({
  name: "ToolbarRevoke",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.revoke,
      "aria-label": t.value.toolbarTips?.revoke,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, Ke);
      },
      type: "button"
    }, [a(x, {
      name: "revoke"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.revoke])]);
  }
}), qo = /* @__PURE__ */ C({
  name: "ToolbarSave",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.save,
      "aria-label": t.value.toolbarTips?.save,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, Ge);
      },
      type: "button"
    }, [a(x, {
      name: "save"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.save])]);
  }
}), Uo = /* @__PURE__ */ C({
  name: "ToolbarStrikeThrough",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.strikeThrough,
      "aria-label": t.value.toolbarTips?.strikeThrough,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "strikeThrough");
      },
      type: "button"
    }, [a(x, {
      name: "strike-through"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.strikeThrough])]);
  }
}), Ko = /* @__PURE__ */ C({
  name: "ToolbarSub",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.sub,
      "aria-label": t.value.toolbarTips?.sub,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "sub");
      },
      type: "button"
    }, [a(x, {
      name: "sub"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.sub])]);
  }
}), Go = /* @__PURE__ */ C({
  name: "ToolbarSup",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.sup,
      "aria-label": t.value.toolbarTips?.sup,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "sup");
      },
      type: "button"
    }, [a(x, {
      name: "sup"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.sup])]);
  }
}), jo = {
  tableShape: {
    type: Array,
    default: () => [6, 4]
  },
  onSelected: {
    type: Function,
    default: () => {
    }
  }
}, Yo = /* @__PURE__ */ C({
  name: "TableShape",
  props: jo,
  setup(e) {
    const t = Y({
      x: -1,
      y: -1
    }), o = P(() => JSON.stringify(e.tableShape)), n = () => {
      const s = [...JSON.parse(o.value)];
      return (!s[2] || s[2] < s[0]) && (s[2] = s[0]), (!s[3] || s[3] < s[3]) && (s[3] = s[1]), s;
    }, i = I(n());
    return N([o], () => {
      i.value = n();
    }), () => a("div", {
      class: `${l}-table-shape`,
      onMouseleave: () => {
        i.value = n(), t.x = -1, t.y = -1;
      }
    }, [new Array(i.value[1]).fill("").map((s, c) => a("div", {
      class: `${l}-table-shape-row`,
      key: `table-shape-row-${c}`
    }, [new Array(i.value[0]).fill("").map((d, u) => a("div", {
      class: `${l}-table-shape-col`,
      key: `table-shape-col-${u}`,
      onMouseenter: () => {
        t.x = c, t.y = u, u + 1 === i.value[0] && u + 1 < i.value[2] ? i.value[0]++ : u + 2 < i.value[0] && i.value[0] > e.tableShape[0] && i.value[0]--, c + 1 === i.value[1] && c + 1 < i.value[3] ? i.value[1]++ : c + 2 < i.value[1] && i.value[1] > e.tableShape[1] && i.value[1]--;
      },
      onClick: () => {
        e.onSelected(t);
      }
    }, [a("div", {
      class: [`${l}-table-shape-col-default`, c <= t.x && u <= t.y && `${l}-table-shape-col-include`]
    }, null)]))]))]);
  }
}), Zo = /* @__PURE__ */ C({
  name: "ToolbarTable",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = r("tableShape"), s = `${e}-toolbar-wrapper`, c = I(!1), d = (m) => {
      c.value = m;
    }, u = (m) => {
      o?.value || h.emit(e, y, "table", {
        selectedShape: m
      });
    }, v = P(() => a(Yo, {
      tableShape: i.value,
      onSelected: u
    }, null));
    return () => a(te, {
      relative: `#${s}`,
      visible: c.value,
      onChange: d,
      disabled: o?.value,
      key: "bar-table",
      overlay: v.value
    }, {
      default: () => [a("button", {
        class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
        title: t.value.toolbarTips?.table,
        "aria-label": t.value.toolbarTips?.table,
        disabled: o?.value,
        type: "button"
      }, [a(x, {
        name: "table"
      }, null), n?.value && a("div", {
        class: `${l}-toolbar-item-name`
      }, [t.value.toolbarTips?.table])])]
    });
  }
}), Xo = /* @__PURE__ */ C({
  name: "ToolbarTask",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.task,
      "aria-label": t.value.toolbarTips?.task,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "task");
      },
      type: "button"
    }, [a(x, {
      name: "task"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.task])]);
  }
}), Jo = /* @__PURE__ */ C({
  name: "ToolbarTitle",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName"), i = `${e}-toolbar-wrapper`, s = I(!1), c = (v) => {
      o?.value || h.emit(e, y, v);
    }, d = (v) => {
      s.value = v;
    }, u = P(() => a("ul", {
      class: `${l}-menu`,
      onClick: () => {
        s.value = !1;
      },
      role: "menu"
    }, [a("li", {
      class: `${l}-menu-item ${l}-menu-item-title`,
      onClick: () => {
        c("h1");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.titleItem?.h1]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-title`,
      onClick: () => {
        c("h2");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.titleItem?.h2]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-title`,
      onClick: () => {
        c("h3");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.titleItem?.h3]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-title`,
      onClick: () => {
        c("h4");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.titleItem?.h4]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-title`,
      onClick: () => {
        c("h5");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.titleItem?.h5]), a("li", {
      class: `${l}-menu-item ${l}-menu-item-title`,
      onClick: () => {
        c("h6");
      },
      role: "menuitem",
      tabindex: "0"
    }, [t.value.titleItem?.h6])]));
    return () => a(te, {
      relative: `#${i}`,
      visible: s.value,
      onChange: d,
      disabled: o?.value,
      overlay: u.value
    }, {
      default: () => [a("button", {
        class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
        disabled: o?.value,
        title: t.value.toolbarTips?.title,
        "aria-label": t.value.toolbarTips?.title,
        type: "button"
      }, [a(x, {
        name: "title"
      }, null), n?.value && a("div", {
        class: `${l}-toolbar-item-name`
      }, [t.value.toolbarTips?.title])])]
    });
  }
}), Qo = /* @__PURE__ */ C({
  name: "ToolbarUnderline",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.underline,
      "aria-label": t.value.toolbarTips?.underline,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "underline");
      },
      type: "button"
    }, [a(x, {
      name: "underline"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.underline])]);
  }
}), ea = /* @__PURE__ */ C({
  name: "ToolbarUnorderedList",
  setup() {
    const e = r("editorId"), t = r("usedLanguageText"), o = r("disabled"), n = r("showToolbarName");
    return () => a("button", {
      class: [`${l}-toolbar-item`, o?.value && `${l}-disabled`],
      title: t.value.toolbarTips?.unorderedList,
      "aria-label": t.value.toolbarTips?.unorderedList,
      disabled: o?.value,
      onClick: () => {
        h.emit(e, y, "unorderedList");
      },
      type: "button"
    }, [a(x, {
      name: "unordered-list"
    }, null), n?.value && a("div", {
      class: `${l}-toolbar-item-name`
    }, [t.value.toolbarTips?.unorderedList])]);
  }
}), ta = () => {
  const e = r("editorId"), t = r("setting"), o = r("updateSetting"), {
    editorExtensions: n,
    editorExtensionsAttrs: i
  } = Z;
  let s = n.screenfull.instance;
  const c = I(!1), d = (m) => {
    if (!s) {
      h.emit(e, X, {
        name: "fullscreen",
        message: "fullscreen is undefined"
      });
      return;
    }
    s.isEnabled ? (c.value = !0, (m === void 0 ? !s.isFullscreen : m) ? s.request() : s.exit()) : console.error("browser does not support screenfull!");
  }, u = () => {
    s && s.isEnabled && s.on("change", () => {
      (c.value || t.value.fullscreen) && (c.value = !1, o("fullscreen"));
    });
  }, v = () => {
    s = window.screenfull, u();
  };
  return K(() => {
    u(), s || Jt("script", {
      ...i.screenfull?.js,
      src: n.screenfull.js,
      id: rt.screenfull,
      onload: v
    }, "screenfull");
  }), K(() => {
    h.on(e, {
      name: Kt,
      callback: d
    });
  }), {
    fullscreenHandler: d
  };
};
let oa = 0;
const Ze = () => {
  const e = r("editorId"), t = r("theme"), o = r("previewTheme"), n = r("language"), i = r("disabled"), s = r("noUploadImg"), c = r("noPrettier"), d = r("codeTheme"), u = r("showToolbarName"), v = r("setting"), m = r("defToolbars");
  return {
    barRender: (g) => {
      if ($t.includes(g))
        switch (g) {
          case "-":
            return a(po, {
              key: `bar-${oa++}`
            }, null);
          case "bold":
            return a(To, {
              key: "bar-bold"
            }, null);
          case "underline":
            return a(Qo, {
              key: "bar-unorderline"
            }, null);
          case "italic":
            return a(Do, {
              key: "bar-italic"
            }, null);
          case "strikeThrough":
            return a(Uo, {
              key: "bar-strikeThrough"
            }, null);
          case "title":
            return a(Jo, {
              key: "bar-title"
            }, null);
          case "sub":
            return a(Ko, {
              key: "bar-sub"
            }, null);
          case "sup":
            return a(Go, {
              key: "bar-sup"
            }, null);
          case "quote":
            return a(_o, {
              key: "bar-quote"
            }, null);
          case "unorderedList":
            return a(ea, {
              key: "bar-unorderedList"
            }, null);
          case "orderedList":
            return a(Po, {
              key: "bar-orderedList"
            }, null);
          case "task":
            return a(Xo, {
              key: "bar-task"
            }, null);
          case "codeRow":
            return a(Co, {
              key: "bar-codeRow"
            }, null);
          case "code":
            return a(yo, {
              key: "bar-code"
            }, null);
          case "link":
            return a(Mo, {
              key: "bar-link"
            }, null);
          case "image":
            return s ? a(So, {
              key: "bar-image"
            }, null) : a(Ao, {
              key: "bar-imageDropdown"
            }, null);
          case "table":
            return a(Zo, {
              key: "bar-table"
            }, null);
          case "revoke":
            return a(zo, {
              key: "bar-revoke"
            }, null);
          case "next":
            return a(Bo, {
              key: "bar-next"
            }, null);
          case "save":
            return a(qo, {
              key: "bar-save"
            }, null);
          case "prettier":
            return !c && a(Oo, {
              key: "bar-prettier"
            }, null);
          case "pageFullscreen":
            return !v.value.fullscreen && a(Ho, {
              key: "bar-pageFullscreen"
            }, null);
          case "fullscreen":
            return a(ko, {
              key: "bar-fullscreen"
            }, null);
          case "catalog":
            return a(wo, {
              key: "bar-catalog"
            }, null);
          case "preview":
            return a(Vo, {
              key: "bar-preview"
            }, null);
          case "previewOnly":
            return a(Wo, {
              key: "bar-previewOnly"
            }, null);
          case "htmlPreview":
            return a(xo, {
              key: "bar-htmlPreview"
            }, null);
          case "github":
            return a($o, {
              key: "bar-github"
            }, null);
          case "mermaid":
            return a(Fo, {
              key: "bar-mermaid"
            }, null);
          case "katex":
            return a(Ro, {
              key: "bar-katex"
            }, null);
        }
      else if (m.value instanceof Array) {
        const p = m.value[g];
        return p ? ne(p, {
          theme: p.props?.theme || t.value,
          previewTheme: p.props?.theme || o.value,
          language: p.props?.theme || n.value,
          codeTheme: p.props?.codeTheme || d.value,
          disabled: p.props?.disabled || i.value,
          showToolbarName: p.props?.showToolbarName || u.value,
          insert(E) {
            h.emit(e, y, "universal", {
              generate: E
            });
          }
        }) : "";
      } else if (m.value?.children instanceof Array) {
        const p = m.value.children[g];
        return p ? ne(p, {
          theme: p.props?.theme || t.value,
          previewTheme: p.props?.theme || o.value,
          language: p.props?.theme || n.value,
          codeTheme: p.props?.codeTheme || d.value,
          disabled: p.props?.disabled || i.value,
          showToolbarName: p.props?.showToolbarName || u.value,
          insert(E) {
            h.emit(e, y, "universal", {
              generate: E
            });
          }
        }) : "";
      } else
        return "";
    }
  };
}, aa = /* @__PURE__ */ C({
  name: "FloatingToolbar",
  setup() {
    const e = r("floatingToolbars"), {
      barRender: t
    } = Ze();
    return () => a("div", {
      class: `${l}-floating-toolbar`
    }, [e.value.map((o) => t(o))]);
  }
}), ge = re.define(), la = qe.define({
  create() {
    return null;
  },
  update(e, t) {
    for (const o of t.effects) o.is(ge) && (e = o.value);
    return e;
  },
  provide: (e) => Vt.from(e)
}), na = (e) => {
  let t = null;
  const o = (s, c) => {
    t && t.kind === c.kind && t.pos === c.pos || (t = c, s.dispatch({
      effects: ge.of({
        pos: c.pos,
        above: !0,
        arrow: !0,
        create: () => {
          const d = document.createElement("div"), u = `${l}-floating-toolbar-container`;
          d.classList.add(u), d.dataset.state = "hidden", requestAnimationFrame(() => {
            d.dataset.state = "visible";
          });
          const v = document.createElement("div");
          d.appendChild(v);
          const m = Ct(aa);
          return e.privide(m), m.mount(d), { dom: d, destroy: () => m.unmount() };
        }
      })
    }));
  }, n = (s) => {
    t && (t = null, s.dispatch({ effects: ge.of(null) }));
  }, i = B.updateListener.of((s) => {
    if (s.selectionSet || s.docChanged) {
      const c = s.state, d = c.selection.main;
      if (!d.empty)
        o(s.view, { kind: "selection", pos: d.anchor });
      else {
        const u = d.head, v = c.doc.lineAt(u);
        /^\s*$/.test(v.text) ? o(s.view, { kind: "emptyLine", pos: u }) : n(s.view);
      }
    }
  });
  return [la, i];
}, ra = "#e5c07b", Le = "var(--md-color)", sa = "#56b6c2", ia = "#fff", Q = "#3f4a54", Ne = "#2d8cf0", ca = "#2d8cf0", ua = "#3f4a54", Ae = "#d19a66", da = "#c678dd", ma = "#f6f6f6", ba = "#ceedfa33", De = "var(--md-bk-color)", de = "var(--md-bk-color)", va = "#bad5fa", Re = "#3f4a54", ga = B.theme(
  {
    "&": {
      color: Q,
      backgroundColor: De
    },
    ".cm-content": {
      caretColor: Re
    },
    ".cm-cursor, .cm-dropCursor": { borderLeftColor: Re },
    "&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": { backgroundColor: va },
    ".cm-panels": { backgroundColor: ma, color: Q },
    ".cm-panels.cm-panels-top": { borderBottom: "1px solid var(--md-border-color)" },
    ".cm-panels.cm-panels-bottom": { borderTop: "1px solid var(--md-border-color)" },
    ".cm-searchMatch": {
      backgroundColor: "#72a1ff59",
      outline: "1px solid #457dff"
    },
    ".cm-searchMatch.cm-searchMatch-selected": {
      backgroundColor: "#6199ff2f"
    },
    ".cm-activeLine": { backgroundColor: "#ceedfa33" },
    ".cm-selectionMatch": { backgroundColor: "#aafe661a" },
    "&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket": {
      backgroundColor: "#bad0f847"
    },
    ".cm-gutters": {
      backgroundColor: De,
      color: Q,
      borderRight: "1px solid",
      borderColor: "var(--md-border-color)"
    },
    ".cm-activeLineGutter": {
      backgroundColor: ba
    },
    ".cm-foldPlaceholder": {
      backgroundColor: "transparent",
      border: "none",
      color: "#ddd"
    },
    ".cm-tooltip": {
      border: "1px solid var(--md-border-color)",
      backgroundColor: de
    },
    ".cm-tooltip .cm-tooltip-arrow:before": {
      borderTopColor: "var(--md-border-color)",
      //'transparent',
      borderBottomColor: "var(--md-border-color)"
    },
    ".cm-tooltip .cm-tooltip-arrow:after": {
      borderTopColor: de,
      borderBottomColor: de
    },
    ".cm-tooltip-autocomplete": {
      "& > ul > li[aria-selected]": {
        color: Q
      }
    }
  }
  // { dark: true }
), ha = Ye.define([
  { tag: b.keyword, color: da },
  { tag: [b.name, b.deleted, b.character, b.propertyName, b.macroName], color: Le },
  { tag: [b.function(b.variableName), b.labelName], color: ca },
  { tag: [b.color, b.constant(b.name), b.standard(b.name)], color: Ae },
  { tag: [b.definition(b.name), b.separator], color: Q },
  {
    tag: [
      b.typeName,
      b.className,
      b.number,
      b.changed,
      b.annotation,
      b.modifier,
      b.self,
      b.namespace
    ],
    color: ra
  },
  {
    tag: [
      b.operator,
      b.operatorKeyword,
      b.url,
      b.escape,
      b.regexp,
      b.link,
      b.special(b.string)
    ],
    color: sa
  },
  { tag: [b.meta, b.comment], color: Ne },
  { tag: b.strong, fontWeight: "bold" },
  { tag: b.emphasis, fontStyle: "italic" },
  { tag: b.strikethrough, textDecoration: "line-through" },
  { tag: b.link, color: Ne, textDecoration: "underline" },
  { tag: b.heading, fontWeight: "bold", color: Le },
  { tag: [b.atom, b.bool, b.special(b.variableName)], color: Ae },
  { tag: [b.processingInstruction, b.string, b.inserted], color: ua },
  { tag: b.invalid, color: ia }
]), Me = [
  ga,
  je(ha)
], fa = "#e5c07b", Fe = "var(--md-color)", pa = "#56b6c2", Ta = "#ffffff", ee = "var(--md-color)", Be = "#e5c07b", wa = "#e5c07b", ya = "var(--md-color)", Pe = "#d19a66", Ca = "#c678dd", ka = "#21252b", $a = "#2c313a", He = "var(--md-bk-color)", me = "var(--md-bk-color)", xa = "#ceedfa33", Oe = "#528bff", Sa = B.theme(
  {
    "&": {
      color: ee,
      backgroundColor: He
    },
    ".cm-content": {
      caretColor: Oe
    },
    ".cm-cursor, .cm-dropCursor": { borderLeftColor: Oe },
    "&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": { backgroundColor: xa },
    ".cm-panels": { backgroundColor: ka, color: ee },
    ".cm-panels.cm-panels-top": { borderBottom: "1px solid var(--md-border-color)" },
    ".cm-panels.cm-panels-bottom": { borderTop: "1px solid var(--md-border-color)" },
    ".cm-searchMatch": {
      backgroundColor: "#72a1ff59",
      outline: "1px solid #457dff"
    },
    ".cm-searchMatch.cm-searchMatch-selected": {
      backgroundColor: "#6199ff2f"
    },
    ".cm-activeLine": { backgroundColor: "#ceedfa33" },
    ".cm-selectionMatch": { backgroundColor: "#aafe661a" },
    "&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket": {
      backgroundColor: "#bad0f847"
    },
    ".cm-gutters": {
      backgroundColor: He,
      color: ee,
      borderRight: "1px solid",
      borderColor: "var(--md-border-color)"
    },
    ".cm-activeLineGutter": {
      backgroundColor: $a
    },
    ".cm-foldPlaceholder": {
      backgroundColor: "transparent",
      border: "none",
      color: "#ddd"
    },
    ".cm-tooltip": {
      border: "1px solid var(--md-border-color)",
      backgroundColor: me
    },
    ".cm-tooltip .cm-tooltip-arrow:before": {
      borderTopColor: "var(--md-border-color)",
      //'transparent',
      borderBottomColor: "var(--md-border-color)"
    },
    ".cm-tooltip .cm-tooltip-arrow:after": {
      borderTopColor: me,
      borderBottomColor: me
    },
    ".cm-tooltip-autocomplete": {
      "& > ul > li[aria-selected]": {
        color: ee
      }
    }
  },
  { dark: !0 }
), Ia = Ye.define([
  { tag: b.keyword, color: Ca },
  { tag: [b.name, b.deleted, b.character, b.propertyName, b.macroName], color: Fe },
  { tag: [b.function(b.variableName), b.labelName], color: wa },
  { tag: [b.color, b.constant(b.name), b.standard(b.name)], color: Pe },
  { tag: [b.definition(b.name), b.separator], color: ee },
  {
    tag: [
      b.typeName,
      b.className,
      b.number,
      b.changed,
      b.annotation,
      b.modifier,
      b.self,
      b.namespace
    ],
    color: fa
  },
  {
    tag: [
      b.operator,
      b.operatorKeyword,
      b.url,
      b.escape,
      b.regexp,
      b.link,
      b.special(b.string)
    ],
    color: pa
  },
  { tag: [b.meta, b.comment], color: Be },
  { tag: b.strong, fontWeight: "bold" },
  { tag: b.emphasis, fontStyle: "italic" },
  { tag: b.strikethrough, textDecoration: "line-through" },
  { tag: b.link, color: Be, textDecoration: "underline" },
  { tag: b.heading, fontWeight: "bold", color: Fe },
  { tag: [b.atom, b.bool, b.special(b.variableName)], color: Pe },
  { tag: [b.processingInstruction, b.string, b.inserted], color: ya },
  { tag: b.invalid, color: Ta }
]), Ve = [
  Sa,
  je(Ia)
], Ea = (e, t) => {
  const o = r("editorId"), n = (s) => {
    s instanceof Promise ? s.then((c) => {
      h.emit(o, y, "universal", {
        generate() {
          return {
            targetValue: c
          };
        }
      });
    }).catch((c) => {
      console.error(c);
    }) : h.emit(o, y, "universal", {
      generate() {
        return {
          targetValue: s
        };
      }
    });
  };
  return (s) => {
    if (!s.clipboardData)
      return;
    if (s.clipboardData.files.length > 0) {
      const { files: g } = s.clipboardData;
      h.emit(
        o,
        fe,
        Array.from(g).filter((p) => /image\/.*/.test(p.type))
      ), s.preventDefault();
      return;
    }
    const c = s.clipboardData.getData("text/plain"), d = t.value?.view.state.selection.main.to || 0, u = t.value?.view.state.doc.lineAt(d).from || 0, v = t.value?.view.state.doc.sliceString(u, d) || "", m = /!\[.*\]\(\s*$/.test(v), f = /!\[.*\]\((.*)\s?.*\)/.test(c);
    if (m) {
      const g = e.transformImgUrl(c);
      n(g), s.preventDefault();
      return;
    } else if (f) {
      const g = c.match(new RegExp(`(?<=!\\[.*\\]\\()([^)\\s]+)(?=\\s?["']?.*["']?\\))`, "g"));
      g ? Promise.all(
        g.map((p) => e.transformImgUrl(p))
      ).then((p) => {
        n(
          p.reduce((k, E, M) => k.replace(g[M], E), c)
        );
      }).catch((p) => {
        console.error(p);
      }) : n(c), s.preventDefault();
      return;
    }
    if (e.autoDetectCode && s.clipboardData.types.includes("vscode-editor-data")) {
      const g = JSON.parse(s.clipboardData.getData("vscode-editor-data"));
      h.emit(o, y, "code", {
        mode: g.mode,
        text: s.clipboardData.getData("text/plain")
      }), s.preventDefault();
      return;
    }
    e.maxlength && c.length + e.modelValue.length > e.maxlength && h.emit(o, X, {
      name: "overlength",
      message: "The input text is too long",
      data: c
    });
  };
}, La = (e, t) => [
  {
    key: "Ctrl-b",
    mac: "Cmd-b",
    run: () => (h.emit(e, y, "bold"), !0)
  },
  {
    key: "Ctrl-d",
    mac: "Cmd-d",
    run: Nt,
    preventDefault: !0
  },
  {
    key: "Ctrl-s",
    mac: "Cmd-s",
    run: (L) => (h.emit(e, Ge, L.state.doc.toString()), !0),
    shift: () => (h.emit(e, y, "strikeThrough"), !0)
  },
  {
    key: "Ctrl-u",
    mac: "Cmd-u",
    preventDefault: !0,
    run: () => (h.emit(e, y, "underline"), !0),
    shift: () => (h.emit(e, y, "unorderedList"), !0)
  },
  {
    key: "Ctrl-i",
    mac: "Cmd-i",
    preventDefault: !0,
    run: () => (h.emit(e, y, "italic"), !0),
    shift: () => (h.emit(e, y, "image"), !0)
  },
  {
    key: "Ctrl-1",
    mac: "Cmd-1",
    run: () => (h.emit(e, y, "h1"), !0)
  },
  {
    key: "Ctrl-2",
    mac: "Cmd-2",
    run: () => (h.emit(e, y, "h2"), !0)
  },
  {
    key: "Ctrl-3",
    mac: "Cmd-3",
    run: () => (h.emit(e, y, "h3"), !0)
  },
  {
    key: "Ctrl-4",
    mac: "Cmd-4",
    run: () => (h.emit(e, y, "h4"), !0)
  },
  {
    key: "Ctrl-5",
    mac: "Cmd-5",
    run: () => (h.emit(e, y, "h5"), !0)
  },
  {
    key: "Ctrl-6",
    mac: "Cmd-6",
    run: () => (h.emit(e, y, "h6"), !0)
  },
  {
    key: "Ctrl-ArrowUp",
    mac: "Cmd-ArrowUp",
    run: () => (h.emit(e, y, "sup"), !0)
  },
  {
    key: "Ctrl-ArrowDown",
    mac: "Cmd-ArrowDown",
    run: () => (h.emit(e, y, "sub"), !0)
  },
  {
    key: "Ctrl-o",
    mac: "Cmd-o",
    run: () => (h.emit(e, y, "orderedList"), !0)
  },
  {
    key: "Ctrl-c",
    mac: "Cmd-c",
    shift: () => (h.emit(e, y, "code"), !0),
    any(L, T) {
      return (T.ctrlKey || T.metaKey) && T.altKey && T.code === "KeyC" ? (h.emit(e, y, "codeRow"), !0) : !1;
    }
  },
  {
    key: "Ctrl-l",
    mac: "Cmd-l",
    run: () => (h.emit(e, y, "link"), !0)
  },
  {
    key: "Ctrl-f",
    mac: "Cmd-f",
    shift: () => t.noPrettier ? !1 : (h.emit(e, y, "prettier"), !0)
  },
  {
    any: (L, T) => (T.ctrlKey || T.metaKey) && T.altKey && T.shiftKey && T.code === "KeyT" ? (h.emit(e, y, "table"), !0) : !1
  },
  ...to
], Na = /[a-z][a-z0-9.+-]*:\/\/[^\s<>"'`()]+(?:\([^\s<>"'`]*\)[^\s<>"'`]*)*/i, Aa = /\/\/[^\s<>"'`()]+/i, Da = /data:[a-z]+\/[a-z0-9.+-]+(?:;base64)?,[a-z0-9+/=%]+/i, Ra = /\/(?!\/)[^\s<>"'`()]+/i, se = new RegExp(
  `(${Na.source}|${Aa.source}|${Da.source}|${Ra.source})`,
  "gi"
), Ma = /[a-z0-9.+-]/i, Fa = (e) => {
  const t = [];
  se.lastIndex = 0;
  let o;
  for (; o = se.exec(e); ) {
    const n = o.index ?? 0, i = n > 0 ? e[n - 1] : "";
    if (i && Ma.test(i) || i === "<" && e[n] === "/")
      continue;
    const c = n + o[0].length;
    t.push([n, c]);
  }
  return t;
}, Ba = (e, t, o) => e.some((n) => n.from === t && n.to === o), Pa = (e) => {
  const t = e.shortenText || (() => "..."), o = re.define(), n = re.define(), i = (u, v) => {
    const m = new Ht(), f = [];
    for (let g = 1; g <= u.doc.lines; g++) {
      const p = u.doc.line(g), k = p.text;
      se.lastIndex = 0;
      const E = e.findTexts?.({
        state: u,
        lineText: k,
        lineNumber: p.number,
        lineFrom: p.from,
        lineTo: p.to,
        defaultTextRegex: se
      }) ?? Fa(k);
      for (const M of E) {
        if (!M) continue;
        const [A, S] = M;
        if (typeof A != "number" || typeof S != "number" || A < 0 || S <= A || A >= k.length || S > k.length)
          continue;
        const D = k.slice(A, S);
        if (!D || D.length <= e.maxLength)
          continue;
        const L = p.from + A, T = p.from + S;
        if (Ba(v, L, T)) {
          f.push({ from: L, to: T });
          continue;
        }
        const R = t(D);
        m.add(
          L,
          T,
          Wt.replace({ widget: new s(R, D, L, T) })
        );
      }
    }
    return { deco: m.finish(), expanded: f };
  };
  class s extends _t {
    constructor(v, m, f, g) {
      super(), this.short = v, this.raw = m, this.from = f, this.to = g;
    }
    toDOM(v) {
      const m = document.createElement("span");
      return m.textContent = this.short, m.className = "cm-short-text", m.title = this.raw, m.style.display = "inline", m.style.textDecoration = "underline", m.addEventListener("mousedown", (f) => {
        f.preventDefault(), f.stopPropagation(), v.dispatch({
          selection: U.cursor(this.from),
          effects: o.of({
            from: this.from,
            to: this.to,
            expand: !0
          })
        }), v.focus();
      }), m.addEventListener("click", (f) => {
        f.preventDefault();
      }), m;
    }
    ignoreEvent() {
      return !1;
    }
    eq(v) {
      return this.short === v.short && this.raw === v.raw && this.from === v.from && this.to === v.to;
    }
  }
  const c = qe.define({
    create(u) {
      return i(u, []);
    },
    update(u, v) {
      let m = u.expanded;
      v.docChanged && m.length && (m = m.map(({ from: g, to: p }) => ({
        from: v.changes.mapPos(g, 1),
        to: v.changes.mapPos(p, -1)
      })).filter(({ from: g, to: p }) => g < p));
      let f = m !== u.expanded;
      for (const g of v.effects)
        g.is(o) ? g.value.expand ? m = [{ from: g.value.from, to: g.value.to }] : m = m.filter(
          ({ from: p, to: k }) => p !== g.value.from || k !== g.value.to
        ) : g.is(n) && m.length > 0 && (m = []);
      return !f && m !== u.expanded && (f = !0), v.docChanged || f ? i(v.state, m) : u;
    },
    provide: (u) => B.decorations.compute([u], (v) => v.field(u).deco)
  }), d = B.domEventHandlers({
    mousedown(u, v) {
      const m = v.state.field(c, !1);
      if (!m || m.expanded.length === 0)
        return !1;
      const f = u.target;
      if (f && v.dom.contains(f)) {
        const g = v.posAtDOM(f, 0);
        if (g != null && g !== -1 && m.expanded.some(
          ({ from: k, to: E }) => g >= k && g <= E
        ))
          return !1;
      }
      return v.dispatch({ effects: n.of(void 0) }), !1;
    }
  });
  return [c, d];
};
B.EDIT_CONTEXT = !1;
const Xe = (e) => e.extension instanceof Function ? e.extension(e.options) : e.extension, Ha = (e) => {
  const t = Xe(e);
  return e.compartment ? e.compartment.of(t) : t;
}, Oa = (e) => {
  const t = r("tabWidth"), o = r("editorId"), n = r("theme"), i = r("previewTheme"), s = r("language"), c = r(
    "usedLanguageText"
  ), d = r("disabled"), u = r("showToolbarName"), v = r("customIcon"), m = r("noUploadImg"), f = r("tableShape"), g = r("noPrettier"), p = r("codeTheme"), k = r("setting"), E = r("updateSetting"), M = r("catalogVisible"), A = r("defToolbars"), S = r("floatingToolbars"), D = r("rootRef"), L = I(), T = kt(), R = I(!1), z = new j(), q = new j(), $ = new j(), Te = new j(), Je = new j(), Qe = La(o, {
    noPrettier: g
  }), we = () => [
    ...Qe,
    ...At,
    ...Dt,
    Rt
  ], oe = {
    paste: Ea(e, T),
    blur: e.onBlur,
    focus: e.onFocus,
    drop: e.onDrop,
    compositionstart: () => {
      R.value = !0;
    },
    compositionend: (w, F) => {
      R.value = !1, e.updateModelValue(F.state.doc.toString());
    },
    input: (w) => {
      e.onInput && e.onInput(w);
      const { data: F } = w;
      e.maxlength && e.modelValue.length + F.length > e.maxlength && h.emit(o, X, {
        name: "overlength",
        message: "The input text is too long",
        data: F
      });
    }
  }, et = na({
    privide(w) {
      w.provide("editorId", o), w.provide("theme", n), w.provide("previewTheme", i), w.provide("language", s), w.provide("disabled", d), w.provide("noUploadImg", m), w.provide("tableShape", f), w.provide("noPrettier", g), w.provide("codeTheme", p), w.provide("showToolbarName", u), w.provide("setting", k), w.provide("updateSetting", E), w.provide("usedLanguageText", c), w.provide("catalogVisible", M), w.provide("defToolbars", A), w.provide("tabWidth", t), w.provide("customIcon", v), w.provide("floatingToolbars", S), w.provide("rootRef", D);
    }
  }), tt = [
    {
      type: "theme",
      extension: ({ theme: w }) => w.value === "light" ? Me : Ve,
      compartment: z,
      options: {
        theme: n
      }
    },
    {
      type: "updateListener",
      extension: B.updateListener.of((w) => {
        w.docChanged && (e.onChange(w.state.doc.toString()), R.value || e.updateModelValue(w.state.doc.toString()));
      })
    },
    {
      type: "domEventHandlers",
      extension: B.domEventHandlers(oe),
      compartment: Te
    },
    {
      type: "completions",
      extension: Ee(e.completions),
      compartment: q
    },
    {
      type: "history",
      extension: $e(),
      compartment: $
    }
  ], ye = Z.codeMirrorExtensions(
    [
      // 横向换行
      {
        type: "lineWrapping",
        extension: B.lineWrapping
      },
      {
        type: "keymap",
        extension: zt.of(we())
      },
      // 解决多行placeholder时，光标异常的情况
      {
        type: "drawSelection",
        extension: qt()
      },
      {
        type: "markdown",
        extension: Bt({ codeLanguages: Pt })
      },
      {
        type: "linkShortener",
        extension: (w) => Pa(w),
        options: {
          maxLength: 30
        }
      },
      {
        type: "floatingToolbar",
        extension: S.value.length > 0 ? et : [],
        compartment: Je
      }
    ],
    {
      editorId: o,
      theme: n.value,
      keyBindings: we()
    }
  ), ot = () => [...tt, ...ye].map(Ha);
  return K(() => {
    const w = new B({
      doc: e.modelValue,
      parent: L.value,
      extensions: ot()
    }), F = new go(w);
    T.value = F, setTimeout(() => {
      F.setTabSize(t), F.setDisabled(d.value), F.setReadOnly(e.readonly), e.placeholder && F.setPlaceholder(e.placeholder), typeof e.maxlength == "number" && F.setMaxLength(e.maxlength), e.autofocus && w.focus();
    }, 0), h.on(o, {
      name: Ke,
      callback() {
        Mt(w);
      }
    }), h.on(o, {
      name: Ue,
      callback() {
        Ft(w);
      }
    }), h.on(o, {
      name: y,
      async callback(O, H = {}) {
        if (O === "image" && H.transform) {
          const V = e.transformImgUrl(H.url);
          if (V instanceof Promise)
            V.then(async (_) => {
              const { text: W, options: J } = await ue(
                O,
                T.value,
                { ...H, url: _ }
              );
              T.value?.replaceSelectedText(W, J, o);
            }).catch((_) => {
              console.error(_);
            });
          else {
            const { text: _, options: W } = await ue(O, T.value, {
              ...H,
              url: V
            });
            T.value?.replaceSelectedText(_, W, o);
          }
        } else {
          const { text: V, options: _ } = await ue(
            O,
            T.value,
            H
          );
          T.value?.replaceSelectedText(V, _, o);
        }
      }
    }), h.on(o, {
      name: Gt,
      callback: Et((O) => {
        const H = { ...oe }, V = Object.keys(oe);
        for (const _ in O) {
          const W = _;
          V.includes(W) ? H[W] = (J, Ce) => {
            O[W](J, Ce), J.defaultPrevented || oe[W](J, Ce);
          } : H[W] = O[W];
        }
        T.value?.view.dispatch({
          effects: Te.reconfigure(
            B.domEventHandlers(H)
          )
        });
      })
    }), h.on(o, {
      name: jt,
      callback: (O, H) => {
        const V = w.state.doc.line(O);
        w.dispatch(
          w.state.update({ changes: { from: V.from, to: V.to, insert: H } })
        );
      }
    }), h.on(o, {
      name: Yt,
      callback() {
        h.emit(o, xe, w);
      }
    }), h.emit(o, xe, w);
  }), N(
    n,
    () => {
      T.value?.view.dispatch({
        effects: z.reconfigure(n.value === "light" ? Me : Ve)
      });
    },
    { deep: !0 }
  ), N(
    () => e.completions,
    () => {
      T.value?.view.dispatch({
        effects: q.reconfigure(Ee(e.completions))
      });
    },
    { deep: !0 }
  ), N(
    () => e.modelValue,
    () => {
      T.value?.getValue() !== e.modelValue && T.value?.setValue(e.modelValue);
    }
  ), N(
    () => e.placeholder,
    () => {
      T.value?.setPlaceholder(e.placeholder);
    }
  ), N([d], () => {
    T.value?.setDisabled(d.value);
  }), N(
    () => e.readonly,
    () => {
      T.value?.setDisabled(e.readonly);
    }
  ), N(
    () => e.maxlength,
    () => {
      e.maxlength && T.value?.setMaxLength(e.maxlength);
    }
  ), fo(() => {
    const w = ye.find(
      (F) => F.type === "floatingToolbar"
    );
    w?.compartment && (S.value.length > 0 ? T.value?.view.dispatch({
      effects: w.compartment.reconfigure(
        Xe(w)
      )
    }) : T.value?.view.dispatch({
      effects: w.compartment.reconfigure([])
    }));
  }, S), {
    inputWrapperRef: L,
    codeMirrorUt: T,
    resetHistory() {
      T.value?.view.dispatch({ effects: $.reconfigure([]) }), T.value?.view.dispatch({ effects: $.reconfigure($e()) });
    }
  };
}, Va = (e, t, o) => {
  const n = r("setting"), i = P(() => /px$/.test(`${e.inputBoxWidth}`) ? "50%" : e.inputBoxWidth), s = Y({
    resizedWidth: i.value
  }), c = Y({
    width: i.value
  }), d = Y({
    insetInlineStart: i.value,
    display: "initial"
  }), u = (f) => {
    const g = t.value?.offsetWidth || 0, p = t.value?.getBoundingClientRect().x || 0;
    let k = f.x - p;
    k / g < ae ? k = g * ae : k > g - g * ae && (k = g - g * ae);
    const E = `${k / g * 100}%`;
    c.width = E, d.insetInlineStart = E, s.resizedWidth = E, e.oninputBoxWidthChange?.(E);
  }, v = (f) => {
    f.target === o.value && document.addEventListener("mousemove", u);
  }, m = () => {
    document.removeEventListener("mousemove", u);
  };
  return N([o], () => {
    document.removeEventListener("mousedown", v), document.removeEventListener("mouseup", m), document.addEventListener("mousedown", v), document.addEventListener("mouseup", m);
  }), K(() => {
    document.addEventListener("mousedown", v), document.addEventListener("mouseup", m);
  }), he(() => {
    document.removeEventListener("mousedown", v), document.removeEventListener("mouseup", m);
  }), N([i], ([f]) => {
    s.resizedWidth = f, c.width = f, d.insetInlineStart = f;
  }), N(
    [
      () => n.value.htmlPreview,
      () => n.value.preview,
      () => n.value.previewOnly
    ],
    () => {
      n.value.previewOnly ? (c.width = "0%", d.display = "none") : !n.value.htmlPreview && !n.value.preview ? (c.width = "100%", d.display = "none") : (c.width = s.resizedWidth, d.display = "initial");
    },
    {
      immediate: !0
    }
  ), {
    inputWrapperStyle: c,
    resizeOperateStyle: d
  };
}, Wa = () => {
  const e = r("editorId"), t = I(!0), o = ze();
  return {
    onCatalogActive: (c, d) => {
      const u = document.querySelector(
        `#${e} .${l}-catalog-editor`
      );
      if (!d || !t.value || !u)
        return;
      const v = d.offsetTop - u.scrollTop;
      (v > 100 || v < 100) && o(u, d.offsetTop - 100);
    },
    onMouseenter: () => t.value = !1,
    onMouseleave: () => t.value = !0
  };
}, be = /* @__PURE__ */ C({
  name: `${xt}CustomScrollbar`,
  props: {
    id: {
      type: String,
      default: void 0
    },
    class: {
      type: [String, Array],
      default: void 0
    },
    /**
     * 滚动元素选择器，默认取 slot 根节点
     */
    scrollTarget: {
      type: String,
      default: void 0
    },
    /**
     * 是否一直显示轨道
     */
    alwaysShowTrack: {
      type: Boolean,
      default: !1
    },
    onMouseenter: {
      type: Function,
      default: () => {
      }
    },
    onMouseleave: {
      type: Function,
      default: () => {
      }
    }
  },
  setup(e, {
    slots: t
  }) {
    const o = I(null), n = I(null), i = I(null), s = I(null);
    let c = null, d = null, u = !1, v = 0, m = 0;
    const f = () => {
      if (!n.value || !o.value || !i.value || !s.value) return;
      const S = o.value.clientHeight, D = n.value.scrollHeight, L = n.value.scrollTop;
      if (D <= S) {
        i.value.style.display = "none", e.alwaysShowTrack || (s.value.style.display = "none");
        return;
      } else
        i.value.style.display = "block", s.value.style.display = "block";
      const T = S / D, R = Math.max(S * T, 20), z = S - R, q = Math.min(L * T, z);
      i.value.style.height = `${R}px`, i.value.style.top = `${q}px`;
    }, g = () => f(), p = (S) => {
      u = !0, v = S.clientY, m = n.value.scrollTop, document.body.style.userSelect = "none";
    }, k = (S) => {
      if (!u || !n.value || !o.value) return;
      const D = S.clientY - v, L = n.value.scrollHeight / o.value.clientHeight;
      n.value.scrollTop = m + D * L;
    }, E = () => {
      u = !1, document.body.style.userSelect = "";
    }, M = (S) => {
      n.value && n.value.removeEventListener("scroll", g), n.value = S, n.value ? (n.value.addEventListener("scroll", g), f()) : s.value && !e.alwaysShowTrack && (s.value.style.display = "none");
    }, A = () => {
      if (!o.value) return;
      const S = e.scrollTarget ? o.value.querySelector(e.scrollTarget) : o.value.firstElementChild;
      M(S);
    };
    return K(async () => {
      await ie(), A(), c = new MutationObserver(() => {
        d && cancelAnimationFrame(d), d = requestAnimationFrame(() => {
          A();
        });
      }), c.observe(o.value, {
        childList: !0,
        subtree: !0
      }), window.addEventListener("resize", f), i.value?.addEventListener("mousedown", p), document.addEventListener("mousemove", k), document.addEventListener("mouseup", E);
    }), he(() => {
      c?.disconnect(), n.value && n.value.removeEventListener("scroll", g), window.removeEventListener("resize", f), i.value?.removeEventListener("mousedown", p), document.removeEventListener("mousemove", k), document.removeEventListener("mouseup", E);
    }), () => a("div", {
      id: e.id,
      class: [`${l}-custom-scrollbar`, e.class],
      ref: o,
      onMouseenter: e.onMouseenter,
      onMouseleave: e.onMouseleave
    }, [t.default?.(), a("div", {
      class: `${l}-custom-scrollbar__track`,
      ref: s
    }, [a("div", {
      class: `${l}-custom-scrollbar__thumb`,
      ref: i
    }, null)])]);
  }
}), _a = ze(), za = {
  flex: 1
}, qa = /* @__PURE__ */ C({
  name: "MDEditorContent",
  props: st,
  setup(e, t) {
    const o = r("editorId"), n = r("catalogVisible"), i = r("theme"), s = r("setting"), c = I(""), d = I(), u = I(), {
      inputWrapperRef: v,
      codeMirrorUt: m,
      resetHistory: f
    } = Oa(e), {
      inputWrapperStyle: g,
      resizeOperateStyle: p
    } = Va(e, d, u);
    oo(e, c, m);
    const {
      onCatalogActive: k,
      onMouseenter: E,
      onMouseleave: M
    } = Wa(), A = (L, T) => {
      if (!s.value.preview && T.line !== void 0) {
        L.preventDefault();
        const R = m.value?.view;
        if (R) {
          const z = R.state.doc.line(T.line + 1), q = R.lineBlockAt(z.from)?.top, $ = R.scrollDOM;
          _a($, q);
        }
      }
    }, S = P(() => s.value.preview ? "preview" : "editor"), D = (L) => {
      c.value = L, e.onHtmlChanged(L);
    };
    return t.expose({
      getSelectedText() {
        return m.value?.getSelectedText();
      },
      focus(L) {
        m.value?.focus(L);
      },
      resetHistory: f,
      getEditorView() {
        return m.value?.view;
      }
    }), () => a("div", {
      class: `${l}-content`
    }, [a("div", {
      class: `${l}-content-wrapper`,
      ref: d
    }, [a(be, {
      alwaysShowTrack: !0,
      scrollTarget: `#${o} .cm-scroller`,
      style: g
    }, {
      default: () => [a("div", {
        class: `${l}-input-wrapper`,
        ref: v
      }, null)]
    }), (s.value.htmlPreview || s.value.preview) && a("div", {
      class: `${l}-resize-operate`,
      style: p,
      ref: u
    }, null), a(be, {
      style: za
    }, {
      default: () => [a(it, {
        modelValue: e.modelValue,
        onChange: e.onChange,
        onHtmlChanged: D,
        onGetCatalog: e.onGetCatalog,
        mdHeadingId: e.mdHeadingId,
        noMermaid: e.noMermaid,
        sanitize: e.sanitize,
        noKatex: e.noKatex,
        formatCopiedText: e.formatCopiedText,
        noHighlight: e.noHighlight,
        noImgZoomIn: e.noImgZoomIn,
        sanitizeMermaid: e.sanitizeMermaid,
        codeFoldable: e.codeFoldable,
        autoFoldThreshold: e.autoFoldThreshold,
        onRemount: e.onRemount,
        previewComponent: e.previewComponent,
        noEcharts: e.noEcharts
      }, null)]
    })]), n.value && a(be, {
      class: `${l}-catalog-${e.catalogLayout}`,
      onMouseenter: E,
      onMouseleave: M
    }, {
      default: () => [a(We, {
        theme: i.value,
        class: `${l}-catalog-editor`,
        editorId: o,
        mdHeadingId: e.mdHeadingId,
        key: "internal-catalog",
        scrollElementOffsetTop: 2,
        syncWith: S.value,
        onClick: A,
        catalogMaxDepth: e.catalogMaxDepth,
        onActive: k
      }, null)]
    })]);
  }
}), Ua = /* @__PURE__ */ C({
  props: {
    modelValue: {
      type: String,
      default: ""
    }
  },
  setup(e) {
    const t = r("usedLanguageText");
    return () => a("div", {
      class: `${l}-footer-item`
    }, [a("label", {
      class: `${l}-footer-label`
    }, [`${t.value.footer?.markdownTotal}:`]), a("span", null, [e.modelValue?.length || 0])]);
  }
}), Ka = {
  checked: {
    type: Boolean,
    default: !1
  },
  onChange: {
    type: Function,
    default: () => {
    }
  },
  disabled: {
    type: Boolean,
    default: void 0
  }
}, Ga = /* @__PURE__ */ C({
  name: `${l}-checkbox`,
  props: Ka,
  setup(e) {
    return () => a("div", {
      class: [`${l}-checkbox`, e.checked && `${l}-checkbox-checked`, e.disabled && `${l}-disabled`],
      onClick: () => {
        e.disabled || e.onChange(!e.checked);
      }
    }, null);
  }
}), ja = {
  scrollAuto: {
    type: Boolean
  },
  onScrollAutoChange: {
    type: Function,
    default: () => {
    }
  }
}, Ya = /* @__PURE__ */ C({
  props: ja,
  setup(e) {
    const t = r("usedLanguageText"), o = r("disabled");
    return () => a("div", {
      class: [`${l}-footer-item`, o?.value && `${l}-disabled`]
    }, [a("label", {
      class: `${l}-footer-label`,
      onClick: () => {
        o?.value || e.onScrollAutoChange(!e.scrollAuto);
      }
    }, [t?.value.footer?.scrollAuto]), a(Ga, {
      checked: e.scrollAuto,
      onChange: e.onScrollAutoChange,
      disabled: o?.value
    }, null)]);
  }
}), Za = {
  modelValue: {
    type: String,
    default: ""
  },
  footers: {
    type: Array,
    default: []
  },
  scrollAuto: {
    type: Boolean
  },
  noScrollAuto: {
    type: Boolean
  },
  onScrollAutoChange: {
    type: Function,
    default: () => {
    }
  },
  defFooters: {
    type: Object
  }
}, Xa = /* @__PURE__ */ C({
  name: "MDEditorFooter",
  props: Za,
  setup(e) {
    const t = r("theme"), o = r("language"), n = r("disabled"), i = P(() => {
      const c = e.footers.indexOf("="), d = c === -1 ? e.footers : e.footers.slice(0, c), u = c === -1 ? [] : e.footers.slice(c, Number.MAX_SAFE_INTEGER);
      return [d, u];
    }), s = (c) => {
      if (St.includes(c))
        switch (c) {
          case "markdownTotal":
            return a(Ua, {
              modelValue: e.modelValue
            }, null);
          case "scrollSwitch":
            return !e.noScrollAuto && a(Ya, {
              scrollAuto: e.scrollAuto,
              onScrollAutoChange: e.onScrollAutoChange
            }, null);
        }
      else if (e.defFooters instanceof Array) {
        const d = e.defFooters[c];
        return d ? ne(d, {
          theme: d.props?.theme || t.value,
          language: d.props?.language || o.value,
          disabled: d.props?.disabled || n?.value
        }) : "";
      } else if (e.defFooters && e.defFooters.children instanceof Array) {
        const d = e.defFooters.children[c];
        return d ? ne(d, {
          theme: d.props?.theme || t.value,
          language: d.props?.language || o.value,
          disabled: d.props?.disabled || n?.value
        }) : "";
      } else
        return "";
    };
    return () => {
      const c = i.value[0].map((u) => s(u)), d = i.value[1].map((u) => s(u));
      return a("div", {
        class: `${l}-footer`
      }, [a("div", {
        class: `${l}-footer-left`
      }, [c]), a("div", {
        class: `${l}-footer-right`
      }, [d])]);
    };
  }
}), Ja = {
  // 工具栏选择显示
  toolbars: {
    type: Array,
    default: () => []
  },
  // 工具栏选择不显示
  toolbarsExclude: {
    type: Array,
    default: () => []
  }
}, Qa = /* @__PURE__ */ C({
  name: "MDEditorToolbar",
  props: Ja,
  setup(e) {
    const t = r("editorId"), o = r("showToolbarName"), n = `${t}-toolbar-wrapper`, i = I(), s = I(), {
      barRender: c
    } = Ze(), d = P(() => {
      const u = e.toolbars.filter((g) => !e.toolbarsExclude.includes(g)), v = u.indexOf("="), m = v === -1 ? u : u.slice(0, v + 1), f = v === -1 ? [] : u.slice(v, Number.MAX_SAFE_INTEGER);
      return [m, f];
    });
    return N(() => e.toolbars, () => {
      ie(() => {
        i.value && Lt(i.value);
      });
    }, {
      immediate: !0
    }), () => {
      const u = d.value[0].map((m) => c(m)), v = d.value[1].map((m) => c(m));
      return a(_e, null, [e.toolbars.length > 0 && a("div", {
        class: `${l}-toolbar-wrapper`,
        ref: i,
        id: n
      }, [a("div", {
        class: [`${l}-toolbar`, o.value && `${l}-stn`]
      }, [a("div", {
        class: `${l}-toolbar-left`,
        ref: s
      }, [u]), a("div", {
        class: `${l}-toolbar-right`
      }, [v])])])]);
    };
  }
}), ve = /* @__PURE__ */ C({
  name: "MdEditorV3",
  props: ut,
  emits: ct,
  setup(e, t) {
    const {
      noKatex: o,
      noMermaid: n,
      noHighlight: i
    } = e, s = Y({
      scrollAuto: e.scrollAuto
    }), c = I(), d = I(), u = P(() => Se({
      props: e,
      ctx: t
    }, "defToolbars")), v = P(() => Se({
      props: e,
      ctx: t
    }, "defFooters")), m = dt(e), [f, g] = mt(e, t, {
      editorId: m
    }), p = bt(e, {
      editorId: m
    });
    vt(e, t, {
      editorId: m
    }), gt(e), ht(e, t, {
      editorId: m
    }), ft(e, t, {
      editorId: m,
      catalogVisible: p,
      setting: f,
      updateSetting: g,
      codeRef: d
    }), pt(e, {
      rootRef: c,
      editorId: m,
      setting: f,
      updateSetting: g,
      catalogVisible: p,
      defToolbars: u
    }), he(() => {
      h.clear(m);
    });
    const k = ($) => {
      t.emit("update:modelValue", $);
    }, E = ($) => {
      e.onChange?.($), t.emit("onChange", $);
    }, M = ($) => {
      e.onHtmlChanged?.($), t.emit("onHtmlChanged", $);
    }, A = ($) => {
      e.onGetCatalog?.($), t.emit("onGetCatalog", $);
    }, S = ($) => {
      e.onBlur?.($), t.emit("onBlur", $);
    }, D = ($) => {
      e.onFocus?.($), t.emit("onFocus", $);
    }, L = ($) => {
      e.onInput?.($), t.emit("onInput", $);
    }, T = ($) => {
      e.onDrop?.($), t.emit("onDrop", $);
    }, R = ($) => {
      e.oninputBoxWidthChange?.($), t.emit("oninputBoxWidthChange", $);
    }, z = () => {
      e.onRemount?.(), t.emit("onRemount");
    }, q = ($) => {
      s.scrollAuto = $;
    };
    return () => a("div", {
      id: m,
      class: [l, e.class, e.theme === "dark" && `${l}-dark`, f.fullscreen || f.pageFullscreen ? `${l}-fullscreen` : ""],
      style: e.style,
      ref: c
    }, [e.toolbars.length > 0 && a(Qa, {
      toolbars: e.toolbars,
      toolbarsExclude: e.toolbarsExclude
    }, null), a(qa, {
      ref: d,
      modelValue: e.modelValue,
      mdHeadingId: e.mdHeadingId,
      noMermaid: n,
      sanitize: e.sanitize,
      placeholder: e.placeholder,
      noKatex: o,
      scrollAuto: s.scrollAuto,
      formatCopiedText: e.formatCopiedText,
      autofocus: e.autoFocus,
      readonly: e.readOnly,
      maxlength: e.maxLength,
      autoDetectCode: e.autoDetectCode,
      noHighlight: i,
      updateModelValue: k,
      onChange: E,
      onHtmlChanged: M,
      onGetCatalog: A,
      onBlur: S,
      onFocus: D,
      onInput: L,
      completions: e.completions,
      noImgZoomIn: e.noImgZoomIn,
      onDrop: T,
      inputBoxWidth: e.inputBoxWidth,
      oninputBoxWidthChange: R,
      sanitizeMermaid: e.sanitizeMermaid,
      transformImgUrl: e.transformImgUrl,
      codeFoldable: e.codeFoldable,
      autoFoldThreshold: e.autoFoldThreshold,
      onRemount: z,
      catalogLayout: e.catalogLayout,
      catalogMaxDepth: e.catalogMaxDepth,
      noEcharts: e.noEcharts,
      previewComponent: e.previewComponent
    }, null), e.footers.length > 0 && a(Xa, {
      modelValue: e.modelValue,
      footers: e.footers,
      defFooters: v.value,
      noScrollAuto: !f.preview && !f.htmlPreview || f.previewOnly,
      scrollAuto: s.scrollAuto,
      onScrollAutoChange: q
    }, null)]);
  }
});
ve.install = (e) => (e.component(ve.name, ve), e.use(yt).use(at).use(wt).use(We).use(Tt), e);
export {
  ve as default
};
